### DISCUSSION TITLE
How do I "promote" my build artifacts?

### DISCUSSION THREAD

**Discussion by rotivleal** [semantic-release/semantic-release#3214] [Q&A]
Inexperienced software dev doing DevOps for the first time.

I am using GitHub Actions.

I want three branches:

1. `alpha` - the trunk, pre-release. every commit in this branch will compile the application, producing build artifacts (Docker images) that will be uploaded to each release.
2. `beta` - pre-release. should _somehow_ react to a new alpha release, download its build artifacts, and test them. if tests pass, release a beta version, retagging the original, now tested, Docker images.
3. `master` - deploy to production.

I am confused on how can I automate `alpha` to `beta` here.

I believe the "semantic-release" way of doing this would be to create a PR from `alpha` to `beta` but:

1. manual? do I really need to create a PR to transfer my commits from `alpha` to `beta` all the time?
2. how would `beta`'s `on push ...` Github workflow know which `alpha` release to download build artifacts for testing from?

GitHub Actions allow you to listen to new releases: https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#release

which would be amazing; I can get the release tag, and use `gh release download <tag>` to download the Docker images for testing.

But for this workflow, the `GITHUB_REF` would not be in `beta`, therefore running `npx semantic-release` in this workflow would maybe just attempt to release to `alpha` again?

I'd love some help because I want to make this work.

Thanks!

**Comment by travi** [comment]
as part of your learning journey, i strongly encourage you to consider if you really want to target a workflow that needs the sort of promotion of changes that you describe. i would suggest that it is likely a more complicated workflow than you truly need. instead, i would strongly recommend following a [trunk-based development](https://minimumcd.org/minimumcd/#trunk-based-development).

additionally, semantic-release is targeted at workflows that resemble trunk-based development rather than those that resemble gitlfow, which seems closer to the workflow you describe. the complications of using semantic-release's pre-release workflow to simulate the overall workflow you describe would be as painful as you describe. this is because that diverges from the intended use of the semantic-release pre-release workflow. it is not intended to enable manual testing of all changes and promotion across multiple channels after testing. it is possible to use it that way, but it is more intended for rare cases where certain changes require more verification than common changes would require in the project's normal workflow.
